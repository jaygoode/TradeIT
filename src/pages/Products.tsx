import React, { useState, useEffect } from "react";
import { useAppSelector, useAppDispatch } from "../hooks/hooks";
import { useNavigate } from "react-router-dom";
import { deleteProductAsync, fetchProducts } from "../reducers/productReducer";
import { addToCart, increaseProductAmount } from "../reducers/cartReducer";
import { ProductsInCart } from "../types/cart";

import ChevronLeftIcon from "@mui/icons-material/ChevronLeft";
import ChevronRightIcon from "@mui/icons-material/ChevronRight";
import {
  IconButton,
  FormControl,
  InputLabel,
  Select,
  MenuItem,
  Button,
  Card,
  CardActions,
  CardContent,
  CardMedia,
  Typography,
  Container,
} from "@mui/material";

const Products = () => {
  const products = useAppSelector((state) => state.productReducer.productList);
  const cartItems = useAppSelector((state) => state.cartReducer);
  const [perPage, setPerPage] = useState(20);
  const [page, setPage] = useState(0);
  const [sorting, setSorting] = useState("price");

  const navigate = useNavigate();
  const dispatch = useAppDispatch();
  const onDelete = (id: string) => {
    dispatch(deleteProductAsync(id));
  };
  const onChange = (input: "left" | "right") => {
    if (page > 0 && input === "left") {
      setPage(page - 1);
    } else {
      setPage(page + 1);
    }
    dispatch(fetchProducts({ offset: page, limit: perPage }));
  };

  const sortingSelect: void = useEffect(() => {
    dispatch(fetchProducts({ offset: page, limit: perPage, sorting: sorting }));
  }, [sorting, perPage]);

  const addToCartHandler = (product: ProductsInCart) => {
    if (cartItems.products.length > 0) {
      console.log(cartItems.products[0].id);
    }

    dispatch(increaseProductAmount(product));
    dispatch(addToCart({ ...product, quantity: 1 }));
  };

  return (
    <>
      <div className="products-container">
        <div className="sorting-container">
          <FormControl className="sorting-form">
            <InputLabel id="sorting-label">Sort by</InputLabel>
            <Select
              labelId="sorting-label"
              id="sorting-select"
              value={sorting}
              label="Age"
              onChange={(e) => setSorting(e.target.value)}
            >
              <MenuItem value="category">Category</MenuItem>
              <MenuItem value="price">Price</MenuItem>
            </Select>
          </FormControl>
          <IconButton onClick={() => onChange("left")}>
            <ChevronLeftIcon />
          </IconButton>
          <FormControl className="perPage-form">
            <InputLabel id="perPage-label">items</InputLabel>
            <Select
              labelId="perPage-label"
              id="perPage-select"
              value={sorting}
              label="items"
              onChange={(e) => setPerPage(Number(e.target.value))}
            >
              <MenuItem value={20}>20 </MenuItem>
              <MenuItem value={60}>60 </MenuItem>
              <MenuItem value={100}>100</MenuItem>
            </Select>
          </FormControl>
          <IconButton onClick={() => onChange("right")}>
            <ChevronRightIcon />
          </IconButton>
        </div>
        <div className="product-section">
          {products.map((product) => (
            <Card key={product.id} className="product">
              <CardMedia
                onClick={() => navigate(`${product.id}`)}
                // className="products__container__img"
                height="200"
                component="img"
                image={product.images[0]}
                alt=""
              />
              <CardContent>
                <Typography>{product.title}</Typography>
                <Typography>${product.price}</Typography>
                <Typography>{product.category.name}</Typography>
              </CardContent>
              <CardActions>
                <Button
                  variant="contained"
                  onClick={() => addToCartHandler({ ...product, quantity: 1 })}
                >
                  Add to cart
                </Button>
                <Button
                  variant="outlined"
                  color="error"
                  onClick={() => onDelete(product.id)}
                >
                  delete
                </Button>
              </CardActions>
            </Card>
          ))}
        </div>

        <div className="bottom-buttons">
          <IconButton onClick={() => onChange("left")}>
            <ChevronLeftIcon />
          </IconButton>
          <FormControl className="perPage-form">
            <InputLabel id="perPage-label">items</InputLabel>
            <Select
              labelId="perPage-label"
              id="perPage-select"
              value={sorting}
              label="items"
              onChange={(e) => setPerPage(Number(e.target.value))}
            >
              <MenuItem value={20}>20 </MenuItem>
              <MenuItem value={60}>60 </MenuItem>
              <MenuItem value={100}>100</MenuItem>
            </Select>
          </FormControl>
          <IconButton onClick={() => onChange("right")}>
            <ChevronRightIcon />
          </IconButton>
        </div>
      </div>
    </>
  );
};

export default Products;
