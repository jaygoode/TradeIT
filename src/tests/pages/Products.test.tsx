import { render } from "../utils/renderer";
import { Renderer } from "react-dom";

import Products from "../../pages/Products";

// describe("testing product page", () => {
//   test("snapshot test", () => {
//     const component = render(<Products />);
//     expect(component).toMatchSnapshot();
//   });

//   test("does Products page have delete button", () => {
//     const { queryAllByText } = render(<Products />);
//     const delbtn = queryAllByText("delete");
//     expect(delbtn).toBeDefined();
//   });
// });
