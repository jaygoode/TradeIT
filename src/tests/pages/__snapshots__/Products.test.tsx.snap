// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing product page snapshot test 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="products__container"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ChevronLeftIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.41 7.41 14 6l-6 6 6 6 1.41-1.41L10.83 12z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <input
          type="number"
        />
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ChevronRightIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="products__container"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ChevronLeftIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.41 7.41 14 6l-6 6 6 6 1.41-1.41L10.83 12z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <input
        type="number"
      />
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ChevronRightIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
